xarray/core/nputils.py

```diff
@@ -129,6 +129,8 @@ def __getitem__(self, key):
 
     def __setitem__(self, key, value):
         """Value must have dimensionality matching the key."""
-        mixed_positions, vindex_positions = _advanced_indexer_subspaces(key)
-        self._array[key] = np.moveaxis(value, vindex_positions,
-                                       mixed_positions)
+        if np.isscalar(value):
```

Comment:
I would suggest using `utils.is_scalar()` and `utils.to_0d_array` for coercing a scalar `value` to a numpy array, then using the original code path here.

Reply:
I just noticed that here we should support broadcasting, rather than some special care of scalars.